app/models/token.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+class Token
+  attr_reader :token_string
+
+  def initialize(token_string)
+    @token_string = token_string
+    validate_token
+  end
+
+  def to_s
+    @token_string
+  end
+
+  def payload
+    @payload ||= if @token_string
+                   pubkey = expected_key
+                   return if pubkey.blank?
+
+                   JWT.decode(@token_string, pubkey, true, algorithm: 'RS256')[0]
+                 end
+  rescue JWT::ExpiredSignature
+    raise error_klass('Validation error: token has expired')
+  rescue JWT::DecodeError
+    raise error_klass('Validation error: token could not be decoded')
+  end
+
+  def expected_key
+    decoded_token = JWT.decode(@token_string, nil, false, algorithm: 'RS256')
+    kid = decoded_token[1]['kid']
+    OIDC::KeyService.get_key(kid)
+  rescue JWT::ExpiredSignature
+    raise error_klass('Validation error: token has expired')
+  rescue JWT::DecodeError
+    raise error_klass('Validation error: token could not be validated')
```

Comment:
Since we have this custom error handling logic for `JWT.decode` could we wrap that in a method so we can deduplicate it (and therefore keep it in sync)?

Reply:
Nevermind, this call to `JWT.decode` doesn't do verification, so I actually think there's no need for these rescues at all.

Reply:
Instead, I think we should just raise a more specific error like "Unable to get `kid` from JWT" or something like that